torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -465,6 +491,60 @@ def _init_reshard_after_forward(self):
                 "sharding_strategy only supports FULL_SHARD and SHARD_GRAD_OP right now."
             )
 
+    def _get_ignored_modules(
+        self,
+        _ignored_modules: Any,
+    ) -> Set[torch.nn.Module]:
+        """
+        Checks that ``_ignored_modules`` is an iterable of ``torch.nn.Module``
+        s without any :class:`FullyShardedDataParallel` instances and then
+        returns them and their children as a :class:`set`, excluding nested
+        :class:`FullyShardedDataParallel` instances.
+
+        We include the child modules to better match user intuition since
+        ignoring a module should ignore its child modules as well, and we
+        exclude :class:`FullyShardedDataParallel` instances since ``self`` may
+        be the intended root instance that manages them.
+        """
+        if _ignored_modules is None:
+            return set()
+        msg_prefix = "`ignored_modules` should be an iterable of " \
+            "`torch.nn.Module`s "
+        try:
+            ignored_root_modules = set(_ignored_modules)
+        except TypeError:
+            raise TypeError(msg_prefix + f"but got {type(_ignored_modules)}")
+        for module in ignored_root_modules:
+            if not isinstance(module, torch.nn.Module):
+                raise TypeError(
+                    msg_prefix + f"but got an iterable with {type(module)}"
+                )
+            if isinstance(module, FullyShardedDataParallel):
+                raise ValueError(
+                    "`ignored_modules` should not include FSDP modules"
+                )
+        # Include child modules and exclude nested FSDP modules
+        ignored_modules = set(
+            child for module in ignored_root_modules
+            for child in module.modules()
+            if not isinstance(child, FullyShardedDataParallel) and
+            not isinstance(child, FlattenParamsWrapper)
+        )
+        return ignored_modules
+
+    def _get_ignored_params(
+        self,
+        ignored_modules: Set[torch.nn.Module],
+    ) -> Set[torch.nn.Parameter]:
+        """
+        Returns the parameters of the modules in ``ignored_modules`` as a
+        :class:`set`, excluding any :class:`FlatParameter` s.
+        """
+        return set(
+            p for m in ignored_modules for p in m.parameters()
+            if not isinstance(p, FlatParameter)
```

Comment:
wondering how we can pass in a ignored_module that contains FlatParameter here?

Reply:
This happens when there is a module in `ignored_modules` that is not `FSDP` but contains a child module that is `FSDP`. This is tested in `test_ignored_modules_nested()` in `test_fsdp_ignored_modules.py`.

This is exactly the case we were debating whether or not to support. On one hand, we can simplify the implementation by disallowing *any* submodules in the modules passed into `ignored_modules` to be `FSDP`, or we can leave the current implementation that allows but skips `FSDP` submodules.